File: apps/bff/src/modules/auth/router.ts
import type { AccessTokenService } from '@bff/modules/auth/access-token';
import type { RefreshTokenStore, UserStore } from '@bff/modules/auth/ports';
import { DOMAIN_ERROR_CODES, DomainError } from '@bff/platform/errors';
import { protectedProcedure, publicProcedure, router, sessionProcedure } from '@bff/trpc/init';
import type { AuthResponse, CurrentUserResponse, RefreshResponse, TokenPair, User } from '@tooday/shared';
import { loginRequestSchema, refreshRequestSchema, signupRequestSchema } from '@tooday/shared';

export interface AuthRouterDeps {
  users: UserStore;
  /** 조립 계층이 제공하는 현재 사용자 조회. 다른 도메인의 구현에 직접 의존하지 않는다. */
  findUserById(userId: string): Promise<User | null>;
  refreshTokens: RefreshTokenStore;
  accessTokens: AccessTokenService;
}

export function createAuthRouter({ users, findUserById, refreshTokens, accessTokens }: AuthRouterDeps) {
  // 로그인/회원가입 공통 — 세션(리프레시) 발급 + 그 세션의 sid로 액세스 서명.
  const issueTokens = async (userId: string): Promise<TokenPair> => {
    const refresh = await refreshTokens.issue(userId);
    return {
      accessToken: await accessTokens.sign({ userId, sessionId: refresh.sessionId }),
      refreshToken: refresh.token,
    };
  };

  return router({
    // 세션 프로브 — 게이트(회원가입/로그인 포함)가 매 진입마다 부른다. 무효 자격증명
    // (만료·폐기) → 401은 sessionProcedure가 거르므로, 여기서 userId가 null이면 자격증명이
    // 아예 없는 익명이다. 익명은 401 에러가 아니라 200 + user:null로 돌려 캐시·dehydrate에
    // 성공으로 남게 한다(재요청 폭주 제거).
    getCurrentUser: sessionProcedure.query(async ({ ctx }): Promise<CurrentUserResponse> => {
      if (!ctx.userId) {
        return { user: null };
      }
      // 액세스 JWT는 userId만 담으므로 전체 프로필은 여기서 지연 조회한다(핫패스가 아니라 허용).
      const user = await findUserById(ctx.userId);
      if (!user) {
        throw new DomainError(DOMAIN_ERROR_CODES.UNAUTHENTICATED);
      }
      return { user };
    }),

    signup: publicProcedure.input(signupRequestSchema).mutation(async ({ ctx, input }): Promise<AuthResponse> => {
      const user = await users.create(input);
      const tokens = await issueTokens(user.id);
      ctx.setAuthCookies(tokens);
      return { user, ...tokens };
    }),

    login: publicProcedure.input(loginRequestSchema).mutation(async ({ ctx, input }): Promise<AuthResponse> => {
      const user = await users.verifyCredentials(input);
      if (!user) {
        throw new DomainError(DOMAIN_ERROR_CODES.INVALID_CREDENTIALS);
      }
      const tokens = await issueTokens(user.id);
      ctx.setAuthCookies(tokens);
      return { user, ...tokens };
    }),

    // 액세스 만료 시 재발급 경로 — 리프레시를 회전(idle 슬라이딩 + absolute 캡)한다.
    // 웹은 리프레시 쿠키로, 네이티브 브릿지는 body로 토큰을 넘긴다.
    refresh: publicProcedure.input(refreshRequestSchema).mutation(async ({ ctx, input }): Promise<RefreshResponse> => {
      const presented = input.refreshToken ?? ctx.refreshToken;
      const rotated = presented ? await refreshTokens.rotate(presented) : null;
      if (!rotated) {
        ctx.clearAuthCookies();
        throw new DomainError(DOMAIN_ERROR_CODES.UNAUTHENTICATED);
      }
      const tokens: TokenPair = {
        accessToken: await accessTokens.sign({ userId: rotated.userId, sessionId: rotated.sessionId }),
        refreshToken: rotated.token,
      };
      ctx.setAuthCookies(tokens);
      return tokens;
    }),

    logout: protectedProcedure.mutation(async ({ ctx }) => {
      if (ctx.refreshToken) {
        await refreshTokens.revoke(ctx.refreshToken);
      }
      ctx.clearAuthCookies();
      return { ok: true };
    }),
  });
}

File: rust/crates/bff/src/modules/auth/router.rs
use std::sync::Arc;

use serde_json::Value;
use tooday_shared::{
    AuthResponse, LoginRequest, LogoutResponse, RefreshRequest, RefreshResponse, SignupRequest, TokenPair,
};

use crate::modules::auth::access_token::{AccessTokenClaims, AccessTokens};
use crate::modules::auth::ports::{CreateUserInput, Credentials, RefreshTokenStore, UserStore};
use crate::platform::errors::{DomainError, DomainErrorCode};
use crate::trpc::context::TrpcContext;
use crate::trpc::init::{output, parse_input, AuthLevel, Procedure, ProcedureKind, TrpcError};

pub const PROCEDURES: &[Procedure] = &[
    Procedure { name: "signup", kind: ProcedureKind::Mutation, auth: AuthLevel::Public },
    Procedure { name: "login", kind: ProcedureKind::Mutation, auth: AuthLevel::Public },
    Procedure { name: "refresh", kind: ProcedureKind::Mutation, auth: AuthLevel::Public },
    Procedure { name: "logout", kind: ProcedureKind::Mutation, auth: AuthLevel::Protected },
];

pub struct AuthRouterDeps {
    pub users: Arc<dyn UserStore>,
    pub refresh_tokens: Arc<dyn RefreshTokenStore>,
    pub access_tokens: Arc<dyn AccessTokens>,
}

impl AuthRouterDeps {
    /// 로그인/회원가입 공통 — 세션(리프레시) 발급 + 그 세션의 sid로 액세스 서명.
    async fn issue_tokens(&self, user_id: &str) -> Result<TokenPair, TrpcError> {
        let refresh = self.refresh_tokens.issue(user_id).await?;
        Ok(TokenPair {
            access_token: self
                .access_tokens
                .sign(&AccessTokenClaims { user_id: user_id.to_owned(), session_id: refresh.session_id }),
            refresh_token: refresh.token,
        })
    }
}

pub async fn call(
    deps: &AuthRouterDeps,
    ctx: &mut TrpcContext,
    name: &str,
    input: Option<Value>,
) -> Result<Value, TrpcError> {
    match name {
        "signup" => {
            let request: SignupRequest = parse_input(input)?;
            let user = deps
                .users
                .create(CreateUserInput {
                    email: request.email,
                    password: request.password,
                    name: request.name,
                })
                .await?;
            let tokens = deps.issue_tokens(&user.id).await?;
            ctx.set_auth_cookies(&tokens);
            output(AuthResponse {
                user,
                access_token: tokens.access_token,
                refresh_token: tokens.refresh_token,
            })
        }

        "login" => {
            let request: LoginRequest = parse_input(input)?;
            let user = deps
                .users
                .verify_credentials(Credentials { email: request.email, password: request.password })
                .await?
                .ok_or_else(|| TrpcError::from(DomainError::new(DomainErrorCode::InvalidCredentials)))?;
            let tokens = deps.issue_tokens(&user.id).await?;
            ctx.set_auth_cookies(&tokens);
            output(AuthResponse {
                user,
                access_token: tokens.access_token,
                refresh_token: tokens.refresh_token,
            })
        }

        // 액세스 만료 시 재발급 경로 — 리프레시를 회전(idle 슬라이딩 + absolute 캡)한다.
        // 웹은 리프레시 쿠키로, 네이티브 브릿지는 body로 토큰을 넘긴다.
        "refresh" => {
            let request: RefreshRequest = parse_input(input)?;
            let presented = request.refresh_token.or_else(|| ctx.refresh_token.clone());
            let rotated = match presented {
                Some(token) => deps.refresh_tokens.rotate(&token).await?,
                None => None,
            };
            let Some(rotated) = rotated else {
                ctx.clear_auth_cookies();
                return Err(DomainError::new(DomainErrorCode::Unauthenticated).into());
            };
            let tokens: RefreshResponse = TokenPair {
                access_token: deps
                    .access_tokens
                    .sign(&AccessTokenClaims { user_id: rotated.user_id, session_id: rotated.session_id }),
                refresh_token: rotated.token,
            };
            ctx.set_auth_cookies(&tokens);
            output(tokens)
        }

        "logout" => {
            if let Some(token) = ctx.refresh_token.clone() {
                deps.refresh_tokens.revoke(&token).await?;
            }
            ctx.clear_auth_cookies();
            output(LogoutResponse { ok: true })
        }

        _ => unreachable!("디스패처가 존재하는 프로시저만 넘긴다"),
    }
}
